File: backend/src/feed.rs
use serde::Serialize;
use serde::ser::{Serializer, SerializeTuple};

use serde_json::to_writer;
use crate::node::Node;
use crate::types::{
    NodeId, NodeStats, NodeHardware, BlockNumber, BlockHash, BlockDetails, Timestamp, Address,
};

pub mod connector;

use connector::Serialized;

pub trait FeedMessage: Serialize {
    const ACTION: u8;
}

pub struct FeedMessageSerializer {
    /// Current buffer,
    buffer: Vec<u8>,
}

impl FeedMessageSerializer {
    pub fn new() -> Self {
        Self {
            buffer: Vec::new(),
        }
    }

    pub fn clear(&mut self) {
        self.buffer.clear();
    }

    pub fn push<Message>(&mut self, msg: Message)
    where
        Message: FeedMessage,
    {
        let glue = match self.buffer.len() {
            0 => b'[',
            _ => b',',
        };

        self.buffer.push(glue);
        let _ = to_writer(&mut self.buffer, &Message::ACTION);
        self.buffer.push(b',');
        let _ = to_writer(&mut self.buffer, &msg);
    }

    pub fn finalize(&mut self) -> Option<Serialized> {
        if self.buffer.len() == 0 {
            return None;
        }

        self.buffer.push(b']');
        let bytes = self.buffer[..].into();
        self.clear();

        Some(Serialized(bytes))
    }
}

macro_rules! actions {
    ($($action:literal: $t:ty,)*) => {
        $(
            impl FeedMessage for $t {
                const ACTION: u8 = $action;
            }
        )*
    }
}

actions! {
    0x00: Version,
    0x01: BestBlock,
    0x02: BestFinalized,
    0x03: AddedNode<'_>,
    0x04: RemovedNode,
    0x05: LocatedNode<'_>,
    0x06: ImportedBlock<'_>,
    0x07: FinalizedBlock,
    0x08: NodeStatsUpdate<'_>,
    0x09: Hardware<'_>,
    0x0A: TimeSync,
    0x0B: AddedChain<'_>,
    0x0C: RemovedChain<'_>,
    0x0D: SubscribedTo<'_>,
    0x0E: UnsubscribedFrom<'_>,
    0x0F: Pong<'_>,
    0x10: AfgFinalized,
    0x11: AfgReceivedPrevote,
    0x12: AfgReceivedPrecommit,
    0x13: AfgAuthoritySet,
    0x14: StaleNode,
}

#[derive(Serialize)]
pub struct Version(pub usize);

#[derive(Serialize)]
pub struct BestBlock(pub BlockNumber, pub Timestamp, pub Option<u64>);

#[derive(Serialize)]
pub struct BestFinalized(pub BlockNumber, pub BlockHash);

pub struct AddedNode<'a>(pub NodeId, pub &'a Node);

#[derive(Serialize)]
pub struct RemovedNode(pub NodeId);

#[derive(Serialize)]
pub struct LocatedNode<'a>(pub NodeId, pub f32, pub f32, pub &'a str);

#[derive(Serialize)]
pub struct ImportedBlock<'a>(pub NodeId, pub &'a BlockDetails);

#[derive(Serialize)]
pub struct FinalizedBlock(pub NodeId, pub BlockNumber, pub BlockHash);

#[derive(Serialize)]
pub struct NodeStatsUpdate<'a>(pub NodeId, pub &'a NodeStats);

#[derive(Serialize)]
pub struct Hardware<'a>(pub NodeId, pub &'a NodeHardware);

#[derive(Serialize)]
pub struct TimeSync(pub u64);

#[derive(Serialize)]
pub struct AddedChain<'a>(pub &'a str, pub usize);

#[derive(Serialize)]
pub struct RemovedChain<'a>(pub &'a str);

#[derive(Serialize)]
pub struct SubscribedTo<'a>(pub &'a str);

#[derive(Serialize)]
pub struct UnsubscribedFrom<'a>(pub &'a str);

#[derive(Serialize)]
pub struct Pong<'a>(pub &'a str);

#[derive(Serialize)]
pub struct AfgFinalized(pub Address, pub BlockNumber, pub BlockHash);

#[derive(Serialize)]
pub struct AfgReceivedPrevote(pub Address, pub BlockNumber, pub BlockHash, pub Option<Address>);

#[derive(Serialize)]
pub struct AfgReceivedPrecommit(pub Address, pub BlockNumber, pub BlockHash, pub Option<Address>);

#[derive(Serialize)]
pub struct AfgAuthoritySet(pub Address, pub Address, pub Address, pub BlockNumber, pub BlockHash);

#[derive(Serialize)]
pub struct StaleNode(pub NodeId);

impl Serialize for AddedNode<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let AddedNode(nid, node) = self;
        let mut tup = serializer.serialize_tuple(7)?;
        tup.serialize_element(nid)?;
        tup.serialize_element(node.details())?;
        tup.serialize_element(node.stats())?;
        tup.serialize_element(node.hardware())?;
        tup.serialize_element(node.block_details())?;
        tup.serialize_element(&node.location())?;
        tup.serialize_element(&node.connected_at())?;
        tup.end()
    }
}

File: frontend/src/common/feed.ts
// Source code for the Substrate Telemetry Server.
// Copyright (C) 2023 Parity Technologies (UK) Ltd.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import { Maybe } from './helpers';
import { stringify, parse, Stringified } from './stringify';
import {
  FeedVersion,
  Address,
  Latitude,
  Longitude,
  City,
  NodeId,
  NodeCount,
  NodeDetails,
  NodeStats,
  NodeIO,
  NodeHardware,
  NodeLocation,
  BlockNumber,
  BlockHash,
  BlockDetails,
  Timestamp,
  Milliseconds,
  ChainLabel,
  GenesisHash,
  AuthoritySetInfo,
  ChainStats,
} from './types';

export const ACTIONS = {
  FeedVersion: 0x00 as const,
  BestBlock: 0x01 as const,
  BestFinalized: 0x02 as const,
  AddedNode: 0x03 as const,
  RemovedNode: 0x04 as const,
  LocatedNode: 0x05 as const,
  ImportedBlock: 0x06 as const,
  FinalizedBlock: 0x07 as const,
  NodeStats: 0x08 as const,
  NodeHardware: 0x09 as const,
  TimeSync: 0x0a as const,
  AddedChain: 0x0b as const,
  RemovedChain: 0x0c as const,
  SubscribedTo: 0x0d as const,
  UnsubscribedFrom: 0x0e as const,
  Pong: 0x0f as const,
  AfgFinalized: 0x10 as const,
  AfgReceivedPrevote: 0x11 as const,
  AfgReceivedPrecommit: 0x12 as const,
  AfgAuthoritySet: 0x13 as const,
  StaleNode: 0x14 as const,
  NodeIO: 0x15 as const,
  ChainStatsUpdate: 0x16 as const,
};

export type Action = typeof ACTIONS[keyof typeof ACTIONS];
export type Payload = Message['payload'];

interface MessageBase {
  action: Action;
}

interface FeedVersionMessage extends MessageBase {
  action: typeof ACTIONS.FeedVersion;
  payload: FeedVersion;
}

interface BestBlockMessage extends MessageBase {
  action: typeof ACTIONS.BestBlock;
  payload: [BlockNumber, Timestamp, Maybe<Milliseconds>];
}

interface BestFinalizedBlockMessage extends MessageBase {
  action: typeof ACTIONS.BestFinalized;
  payload: [BlockNumber, BlockHash];
}

interface AddedNodeMessage extends MessageBase {
  action: typeof ACTIONS.AddedNode;
  payload: [
    NodeId,
    NodeDetails,
    NodeStats,
    NodeIO,
    NodeHardware,
    BlockDetails,
    Maybe<NodeLocation>,
    Maybe<Timestamp>
  ];
}

interface RemovedNodeMessage extends MessageBase {
  action: typeof ACTIONS.RemovedNode;
  payload: NodeId;
}

interface LocatedNodeMessage extends MessageBase {
  action: typeof ACTIONS.LocatedNode;
  payload: [NodeId, Latitude, Longitude, City];
}

interface ImportedBlockMessage extends MessageBase {
  action: typeof ACTIONS.ImportedBlock;
  payload: [NodeId, BlockDetails];
}

interface FinalizedBlockMessage extends MessageBase {
  action: typeof ACTIONS.FinalizedBlock;
  payload: [NodeId, BlockNumber, BlockHash];
}

interface NodeStatsMessage extends MessageBase {
  action: typeof ACTIONS.NodeStats;
  payload: [NodeId, NodeStats];
}

interface NodeHardwareMessage extends MessageBase {
  action: typeof ACTIONS.NodeHardware;
  payload: [NodeId, NodeHardware];
}

interface NodeIOMessage extends MessageBase {
  action: typeof ACTIONS.NodeIO;
  payload: [NodeId, NodeIO];
}

interface TimeSyncMessage extends MessageBase {
  action: typeof ACTIONS.TimeSync;
  payload: Timestamp;
}

interface AddedChainMessage extends MessageBase {
  action: typeof ACTIONS.AddedChain;
  payload: [ChainLabel, GenesisHash, NodeCount];
}

interface RemovedChainMessage extends MessageBase {
  action: typeof ACTIONS.RemovedChain;
  payload: GenesisHash;
}

interface SubscribedToMessage extends MessageBase {
  action: typeof ACTIONS.SubscribedTo;
  payload: GenesisHash;
}

interface UnsubscribedFromMessage extends MessageBase {
  action: typeof ACTIONS.UnsubscribedFrom;
  payload: GenesisHash;
}

interface PongMessage extends MessageBase {
  action: typeof ACTIONS.Pong;
  payload: string; // just echo whatever `ping` sent
}

interface AfgFinalizedMessage extends MessageBase {
  action: typeof ACTIONS.AfgFinalized;
  payload: [Address, BlockNumber, BlockHash];
}

interface AfgAuthoritySet extends MessageBase {
  action: typeof ACTIONS.AfgAuthoritySet;
  payload: AuthoritySetInfo;
}

interface AfgReceivedPrecommit extends MessageBase {
  action: typeof ACTIONS.AfgReceivedPrecommit;
  payload: [Address, BlockNumber, BlockHash, Address];
}

interface AfgReceivedPrevote extends MessageBase {
  action: typeof ACTIONS.AfgReceivedPrevote;
  payload: [Address, BlockNumber, BlockHash, Address];
}

interface StaleNodeMessage extends MessageBase {
  action: typeof ACTIONS.StaleNode;
  payload: NodeId;
}

interface ChainStatsUpdate extends MessageBase {
  action: typeof ACTIONS.ChainStatsUpdate;
  payload: ChainStats;
}

export type Message =
  | FeedVersionMessage
  | BestBlockMessage
  | BestFinalizedBlockMessage
  | AddedNodeMessage
  | RemovedNodeMessage
  | LocatedNodeMessage
  | ImportedBlockMessage
  | FinalizedBlockMessage
  | NodeStatsMessage
  | NodeHardwareMessage
  | TimeSyncMessage
  | AddedChainMessage
  | RemovedChainMessage
  | SubscribedToMessage
  | UnsubscribedFromMessage
  | AfgFinalizedMessage
  | AfgReceivedPrevote
  | AfgReceivedPrecommit
  | AfgAuthoritySet
  | StaleNodeMessage
  | PongMessage
  | NodeIOMessage
  | ChainStatsUpdate;

/**
 * Data type to be sent to the feed. Passing through strings means we can only serialize once,
 * no matter how many feed clients are listening in.
 */
export type SquashedMessages = Array<Action | Payload>;
export type Data = Stringified<SquashedMessages>;

/**
 * Serialize an array of `Message`s to a single JSON string.
 *
 * All messages are squashed into a single array of alternating opcodes and payloads.
 *
 * Action `string`s are converted to opcodes using the `actionToCode` mapping.
 */
export function serialize(messages: Array<Message>): Data {
  const squashed: SquashedMessages = new Array(messages.length * 2);
  let index = 0;

  messages.forEach((message) => {
    const { action, payload } = message;

    squashed[index++] = action;
    squashed[index++] = payload;
  });

  return stringify(squashed);
}

/**
 * Deserialize data to an array of `Message`s.
 */
export function deserialize(data: Data): Array<Message> {
  const json = parse(data);

  if (!Array.isArray(json) || json.length === 0 || json.length % 2 !== 0) {
    throw new Error('Invalid FeedMessage.Data');
  }

  const messages = new Array<Message>(json.length / 2);

  for (const index of messages.keys()) {
    const [action, payload] = json.slice(index * 2);

    messages[index] = { action, payload } as Message;
  }

  return messages;
}
